packages/shell-api/src/deep-inspect-service-provider-wrapper.ts

```diff
@@ -0,0 +1,257 @@
+      cursor.tryNext.bind(cursor) as () => Promise<Document | null>
+    );
+
+    if (cursor.readBufferedDocuments) {
+      cursor.readBufferedDocuments = cursorReadBufferedDocuments(
+        cursor.readBufferedDocuments.bind(cursor) as (
+          number?: number
+        ) => Document[]
+      );
+    }
+    if (cursor.toArray) {
+      cursor.toArray = cursorToArray(
+        cursor.toArray.bind(cursor) as () => Promise<Document[]>
+      );
+    }
+
+    return cursor;
+  };
+}
+
+const customInspectSymbol = Symbol.for('nodejs.util.inspect.custom');
+
+function cursorNext(
+  original: () => Promise<Document | null>
+): () => Promise<Document | null> {
+  return async function (): Promise<Document | null> {
+    const result = await original();
+    if (result) {
+      replaceWithCustomInspect(result);
+    }
+    return result;
+  };
+}
+
+const cursorTryNext = cursorNext;
+
+function cursorReadBufferedDocuments(
+  original: (number?: number) => Document[]
+): (number?: number) => Document[] {
+  return function (number?: number): Document[] {
+    const results = original(number);
+
+    replaceWithCustomInspect(results);
+
+    return results;
+  };
+}
+
+function cursorToArray(
+  original: () => Promise<Document[]>
+): () => Promise<Document[]> {
+  return async function (): Promise<Document[]> {
+    const results = await original();
+
+    replaceWithCustomInspect(results);
+
+    return results;
+  };
+}
+
+function bsonMethod<
+  K extends keyof PickMethodsByReturnType<ServiceProvider, Promise<any>>
+>(
+  key: K
+): (
+  ...args: Parameters<Required<ServiceProvider>[K]>
+) => ReturnType<Required<ServiceProvider>[K]> {
+  return async function (
+    this: DeepInspectServiceProviderWrapper,
+    ...args: Parameters<Required<ServiceProvider>[K]>
+  ): // eslint-disable-next-line @typescript-eslint/ban-ts-comment
+  // @ts-ignore The returntype already contains a promise
+  ReturnType<Required<ServiceProvider>[K]> {
+    const result = await (this._sp[key] as any)(...args);
+    replaceWithCustomInspect(result);
+    return result;
+  };
+}
+
+function forwardedMethod<
+  K extends keyof PickMethodsByReturnType<ServiceProvider, any>
+>(
+  key: K
+): (
+  ...args: Parameters<Required<ServiceProvider>[K]>
+) => ReturnType<Required<ServiceProvider>[K]> {
+  return function (
+    this: DeepInspectServiceProviderWrapper,
+    ...args: Parameters<Required<ServiceProvider>[K]>
+  ): ReturnType<Required<ServiceProvider>[K]> {
+    // not wrapping the result at all because forwardedMethod() is for simple
+    // values only
+    return (this._sp[key] as any)(...args);
+  };
+}
+
+function customDocumentInspect(
+  this: Document,
+  depth: number,
+  inspectOptions: InspectOptions,
+  inspect: typeof _inspect
+) {
+  const newInspectOptions = {
+    ...inspectOptions,
+    depth: Infinity,
+    maxArrayLength: Infinity,
+    maxStringLength: Infinity,
+  };
+
+  // reuse the standard inpect logic for an object without causing infinite
```

Comment:
This feels very hacky, but I haven't thought of a better way to do it yet. I'm trying to avoid having to write my own function that will format an array or object exactly the way util.inspect() already does.